hs_core/hydroshare/features.py

```diff
@@ -0,0 +1,475 @@
+from django.contrib.auth.models import User, Group
+from hs_access_control.models import PrivilegeCodes, GroupAccess, GroupResourcePrivilege
+from hs_core.models import BaseResource
+from hs_core.search_indexes import BaseResourceIndex
+from hs_tracking.models import Variable
+import re
+
+
+class Features(object):
+    """
+    Features appropriate for machine learning analysis of our databases.
+    """
+
+    @staticmethod
+    def resource_owners():
+        """ return (username, resource_id) tuples representing resource owners """
+        records = []
+        for u in User.objects.all():
+            owned = BaseResource.objects.filter(r2urp__user=u,
+                                                r2urp__privilege=PrivilegeCodes.OWNER)
+            for r in owned:
+                records.append((u.username, r.short_id,))
+        return records
+
+    @staticmethod
+    def group_owners():
+        """ return (username, group_name) tuples representing group ownership """
+        records = []
+        for u in User.objects.all():
+            owned = Group.objects.filter(g2ugp__user=u,
+                                         g2ugp__privilege=PrivilegeCodes.OWNER)
+            for g in owned:
+                records.append((u.username, g.name,))
+        return records
+
+    @staticmethod
+    def resource_editors():
+        """ return (username, resource_id) tuples representing resource editing privilege """
+        records = []
+        for u in User.objects.all():
+            editable = BaseResource.objects.filter(r2urp__user=u,
+                                                   r2urp__privilege__lte=PrivilegeCodes.CHANGE)
+            for r in editable:
+                records.append((u.username, r.short_id,))
+        return records
+
+    @staticmethod
+    def group_editors():
+        """ return (username, group_name) tuples representing group editing privilege """
+        records = []
+        for u in User.objects.all():
+            editable = Group.objects.filter(g2ugp__user=u,
+                                            g2ugp__privilege__lte=PrivilegeCodes.CHANGE)
+            for g in editable:
+                records.append((u.username, g.name,))
+        return records
+
+    @staticmethod
+    def resource_viewers(fromdate, todate):
```

Comment:
Same with these resource lists from a date range.  The example I did above may help you adapt this to a query too.  Let me know if you need help with it.